### ISSUE TITLE
What is the best way to add an attribute to an object in JavaScript?

### ISSUE THREAD

**Issue by acbart** [skulpt/skulpt#732]
If I am creating a third party library (e.g., my own Requests module), and it creates a class, what is the best way to add attributes/fields to that class? I see many examples of how to add methods, but I'm still unclear about the fields. I've seen a few ways:

* Sk.abstr.sattr
* Implement \_\_getattr__

Is there a preferred, recommended way?

**Comment by albertjan** [comment]
I hope you kept an eye on the request module in the stdlib :) https://github.com/skulpt/skulpt/blob/master/src/lib/urllib/request/__init__.js. 

If you have to run a function to get the value of your variable I usually go with a `__getattr__` function with a switch statement on the keys.

But if it's just a variable I would just `sattr` it onto self in `__init__`. 
`Sk.abstr.sattr(self, 'var_name', new Sk.builtin.int_(5))` that will use the generic get and set attr functions in class.

**Comment by acbart** [comment]
Yep, stole a lot of the implementation from Urllib, it was an extremely helpful base. I mostly just had to add some new attributes and allow it to use my own custom function to do the actual requests (since I need to send them through a proxy server - which is actually usually returning cached results anyway!).

**Comment by acbart** [comment]
I was about to open a new issue about this, when I discovered that I had already asked this question before. However, times have changed code and I was hoping to get some clarification.

The new version of Skulpt now does some additional work when compiling the `attr` part of Attribute nodes. In addition to mangling, fixing reserved words, and fixing reserved names, the system generates a constant for the string value. This string value gets accessed consistently when ever this `attr` gets referenced, so there are no issues.

If you are simply calling `Sk.abstr.gattr` or `Sk.abstr.sattr` without doing this work, as you would in a custom JS module, it seems like it is possible for a name to not get mangled coming in but does get mangled when regularly compiled code does. This is particularly a problem for my Ast library that needs to have nodes with attributes that are reserved names (e.g., `FunctionDef` has a `name` attribute). So I set the attribute (as `name`), but cannot get it back (because user code will generate `name_$rn$`).

Is there an updated recommendation for using `sattr` in our custom modules, perhaps to do this mangling myself? Or should I just implement `__getattr__` from now on?
